sdk/identity/azure-identity/azure/identity/aio/_internal.py

```diff
@@ -4,85 +4,102 @@
 # license information.
 # ------------------------------------------------------------------------
 import os
-from typing import Any, Dict, Optional
+from typing import Any, Optional
 
 from azure.core import Configuration
 from azure.core.credentials import AccessToken
 from azure.core.pipeline.policies import ContentDecodePolicy, HeadersPolicy, NetworkTraceLoggingPolicy, AsyncRetryPolicy
+from urllib3.connection import ConnectTimeoutError  # urllib3 should be available because requests depends on it
 
 from ._authn_client import AsyncAuthnClient
 from ..constants import Endpoints, MSI_ENDPOINT, MSI_SECRET
 from ..exceptions import AuthenticationError
+from .._internal import _ManagedIdentityBase
 
 
-class AsyncImdsCredential:
-    def __init__(self, config: Optional[Configuration] = None, **kwargs: Dict[str, Any]) -> None:
-        config = config or self.create_config(**kwargs)
-        policies = [config.header_policy, ContentDecodePolicy(), config.retry_policy, config.logging_policy]
-        self._client = AsyncAuthnClient(Endpoints.IMDS, config, policies, **kwargs)
+class _AsyncManagedIdentityBase(_ManagedIdentityBase):
+    def __init__(self, endpoint: str, config: Optional[Configuration] = None, **kwargs: Any) -> None:
+        super().__init__(endpoint=endpoint, config=config, client_cls=AsyncAuthnClient, **kwargs)
+
+    @staticmethod
+    def create_config(**kwargs: Any) -> Configuration:  # type: ignore
+        return _ManagedIdentityBase.create_config(retry_policy=AsyncRetryPolicy, **kwargs)
+
+
+class AsyncImdsCredential(_AsyncManagedIdentityBase):
+    def __init__(self, config: Optional[Configuration] = None, **kwargs: Any) -> None:
+        super().__init__(endpoint=Endpoints.IMDS, config=config, **kwargs)
+        self._endpoint_available = None  # type: Optional[bool]
 
     async def get_token(self, *scopes: str) -> AccessToken:
+        if self._endpoint_available is None:
+            # Lacking another way to determine whether the IMDS endpoint is listening,
+            # we send a request it would immediately reject (missing a required header),
+            # setting a short timeout.
+            try:
+                await self._client.request_token(scopes, method="GET", connection_timeout=0.3)
+                self._endpoint_available = True
+            except AuthenticationError:
+                # received a response that couldn't be deserialized (because it was an error response)
+                self._endpoint_available = True
+            except ConnectTimeoutError:
+                self._endpoint_available = False
```

Comment:
Are there other possible error types besides AuthenticationError and ConnectTimeoutError possible here?  Should we catch more broadly for the case when we set _enpoint_available to False?

Reply:
Changed this to consider HttpResponseError (got a response, some policy choked on it) and AuthenticationError (got a response, AuthnClient choked on it) as indicating availability, anything else as indicating unavailability.